pkg/ccl/backupccl/restore_online.go

```diff
@@ -230,3 +242,171 @@ func checkRewritesAreNoops(rewrites jobspb.DescRewriteMap) error {
+	execCtx sql.JobExecContext, downloadSpansCh chan roachpb.Span,
+) func(context.Context) error {
+	return func(ctx context.Context) error {
+		ctx, tsp := tracing.ChildSpan(ctx, "backupccl.sendDownloadWorker")
+		defer tsp.Finish()
+		for sp := range downloadSpansCh {
+			log.VInfof(ctx, 1, "sending download request for span %s", sp)
+			var resp *serverpb.DownloadSpanResponse
+			var err error
+			if resp, err = execCtx.ExecCfg().TenantStatusServer.DownloadSpan(ctx, &serverpb.DownloadSpanRequest{
+				Span: sp,
+			}); err != nil {
+				return err
+			}
+			if len(resp.ErrorsByNodeID) > 0 {
+				return errors.Newf("failed to download span %s on all nodes: %v", sp, resp.ErrorsByNodeID)
+			}
+		}
+		return nil
+	}
+}
+
+// waitForDownloadToComplete waits until there are no more ExternalFileBytes
+// remaining to be downloaded for the restore.
+func (r *restoreResumer) waitForDownloadToComplete(
+	ctx context.Context, execCtx sql.JobExecContext, details jobspb.RestoreDetails,
+) error {
+	ctx, tsp := tracing.ChildSpan(ctx, "backupccl.waitForDownloadToComplete")
+	defer tsp.Finish()
+	total, err := r.maybeCalculateTotalDownloadSpans(ctx, execCtx, details)
+	if err != nil {
+		return errors.Wrap(err, "failed to calculate total number of spans to download")
+	}
+
+	// Download is already complete or there is nothing to be downloaded, in
+	// either case we can mark the job as done.
+	if total == 0 {
+		if err := r.job.NoTxn().FractionProgressed(ctx, func(ctx context.Context, details jobspb.ProgressDetails) float32 {
+			return 1.0
+		}); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	var lastProgressUpdate time.Time
+	for rt := retry.StartWithCtx(
+		ctx, retry.Options{InitialBackoff: time.Second, MaxBackoff: time.Second * 10},
+	); ; rt.Next() {
+
+		var remaining uint64
+		for _, span := range details.DownloadSpans {
+			resp, err := execCtx.ExecCfg().TenantStatusServer.SpanStats(ctx, &roachpb.SpanStatsRequest{
+				Spans: []roachpb.Span{span},
+			})
+			if err != nil {
+				return err
+			}
+			for _, stats := range resp.SpanToStats {
+				remaining += stats.ExternalFileBytes
+			}
+		}
+
+		fractionComplete := float32(total-remaining) / float32(total)
+		log.Infof(ctx, "restore download phase, %s downloaded, %s remaining of %s total (%.2f complete)",
+			sz(total-remaining), sz(remaining), sz(total), fractionComplete,
+		)
+
+		if remaining == 0 {
+			return nil
+		}
+
+		if timeutil.Since(lastProgressUpdate) > time.Minute {
+			if err := r.job.NoTxn().FractionProgressed(ctx, func(ctx context.Context, details jobspb.ProgressDetails) float32 {
+				return fractionComplete
+			}); err != nil {
+				return err
+			}
+			lastProgressUpdate = timeutil.Now()
+		}
+	}
+}
+
+func (r *restoreResumer) doDownloadFiles(ctx context.Context, execCtx sql.JobExecContext) error {
+	details := r.job.Details().(jobspb.RestoreDetails)
+
+	grp := ctxgroup.WithContext(ctx)
+	downloadSpansCh := make(chan roachpb.Span, len(details.DownloadSpans))
```

Comment:
I didn't want this channel to be a bottleneck. We could have multiple workers pulling off this channel and issuing download requests, and we should just fill the channel as quickly as we can to avoid any of those workers waiting for work.

Reply:
I see. I don't have strong opinions here, so I'm fine to keep it as is.